test-case-reporting/src/test_result_record.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright 2020 The AMP HTML Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Build, DB, Job, TestCase, TestRun} from 'test-case-reporting';
+import {Database, TIMESTAMP_PRECISION} from './db';
+
+const msConversionConstant: number = Math.pow(10, 3 - TIMESTAMP_PRECISION);
+
+// Types in the DB namespace interface use snake_case instead of camelCase.
+/* eslint @typescript-eslint/camelcase: "off" */
+
+function getDateFromTimestamp(timestamp: number): Date {
+  return new Date(timestamp * msConversionConstant);
+}
+
+export class TestResultRecord {
+  constructor(private db: Database) {}
+
+  /**
+   * Gets a DB.Build from the database by its build number.
+   * If there is no such build, it returns undefined.
+   * @param buildNumber The number of the Travis build, as a string.
+   */
+  private async getDbBuildFromBuildNumber(
+    buildNumber: string
+  ): Promise<DB.Build> {
+    return this.db<DB.Build>('builds')
+      .where('build_number', buildNumber)
+      .first();
+  }
+
+  /**
+   * Makes a Build from a DB.Build.
+   * @param dbBuild A build as obtained from the database.
+   */
+  private getBuildFromDbBuild(dbBuild: DB.Build): Build {
+    return {
+      commitSha: dbBuild.commit_sha,
+      buildNumber: dbBuild.build_number,
+      startedAt: getDateFromTimestamp(dbBuild.started_at),
+    };
+  }
+
+  /**
+   * Gets a Build from the database by its build number.
+   * If there is no such build, it returns undefined.
+   * @param buildNumber The number of the Travis build, as a string
+   */
+  private async getBuildFromBuildNumber(buildNumber: string): Promise<Build> {
+    const dbBuild = await this.getDbBuildFromBuildNumber(buildNumber);
+    return this.getBuildFromDbBuild(dbBuild);
+  }
+
+  private async getTestRunsOfBuild(
+    buildNumber: string
+  ): Promise<Array<TestRun>> {
+    const build: Build = await this.getBuildFromBuildNumber(buildNumber);
+    const dbJobs: Array<DB.Job> = await this.db<DB.Job>('builds')
```

Comment:
I suspect this query can also JOIN against the other tables and include the job, test run, and test case info all in one query.

Reply:
A benefit of the current approach is that it creates only one `Job` per job -- i.e. multiple `TestRun`s that belong to the same `Job` actually reference the same object. If we did a larger join that included job, test case, and rest run info, I don't think I could do the same thing. In your experience, is this kind of benefit worth it, or not really?

Reply:
What benefits are gained by sharing a single Job object? How does it impact the way the results will be used?

Reply:
It doesn't impact the way the results will be used, but it avoids duplication of `Job` objects and makes it _slightly_ simpler to check whether two test runs belong to the same job. It also feels structurally cleaner; though I'll acknowledge that's a subjective aesthetic preference and not super practical.

I'm guessing from the conversation that the benefits aren't worth the added database queries -- let me know if my guess is correct. I'll be working on a single-join approach in the meantime.

Reply:
Done implementing single-query approach.